From 70fcc05761a0a9614c041d4b0517bb41ea5f9ffe Mon Sep 17 00:00:00 2001
From: Radha Mohan Chintakuntla <radhac@marvell.com>
Date: Wed, 29 Aug 2018 14:29:06 -0700
Subject: [PATCH 0446/1921] gpio: thunderx: Remove char device when exiting
 driver

Previous patch "gpio: thunderx: Add support for EL0 interrupts for GPIO"
was missing the cleanup part that removes the special character device
when the driver is exiting.

Change-Id: I34bab9d12407b7fc8d315d62908e1d08056af1fb
Signed-off-by: Radha Mohan Chintakuntla <radhac@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/8767
Reviewed-by: Sujeet Kumar Baranwal <Sujeet.Baranwal@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26941
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/gpio/gpio-thunderx.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpio/gpio-thunderx.c b/drivers/gpio/gpio-thunderx.c
index 7d99a43b3ada..dd7ed74e4397 100644
--- a/drivers/gpio/gpio-thunderx.c
+++ b/drivers/gpio/gpio-thunderx.c
@@ -865,6 +865,15 @@ static void thunderx_gpio_remove(struct pci_dev *pdev)
 	irq_domain_remove(txgpio->irqd);
 
 	pci_set_drvdata(pdev, NULL);
+
+#ifdef CONFIG_MRVL_OCTEONTX_EL0_INTR
+	device_destroy(otx_class, otx_dev);
+	class_destroy(otx_class);
+	cdev_del(otx_cdev);
+	unregister_chrdev_region(otx_dev, 1);
+
+	task_cleanup_handler_remove(cleanup_el3_irqs);
+#endif
 }
 
 static const struct pci_device_id thunderx_gpio_id_table[] = {
-- 
2.31.1

